Playwright: implement a basic browser manager.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

Implement a basic `browser-manager.js` that provides convenient utilities to manage the Playwright browser instance. Its functions are modelled after the Selenium-based `driver-helper.js` file.

##### Functions implemented

- screen size
- locale
- generate new context
- launch new browser
- quit browser

Note that:
- BrowserManager uses BrowserContexts instead of starting a new Browser instance for [performance reasons](https://playwright.dev/docs/core-concepts#browser-contexts). 
- Only one browser is ever launched.

The demonstration script `Auth screen canary` has been updated to use abilities offered by browser-manager.

#### Testing instructions

Ensure that tests can run on TeamCity using the new browser manager. Example TeamCity run is shown [here](https://teamcity.a8c.com/buildConfiguration/calypso_RunCalypsoPlaywrightE2eTests/6019216).

Partially implements #51082.

worldomonation committed:
Implement a basic form of browser manager file for Playwright.
- implement ability to specify launch parameters for Browser instance.
- implement ability to generate new instances of BrowserContexts with minimal boilerplate code in the tests.
- implement other helper functions to lay the groundwork for future work on screenshot and video capture.

Use the new BrowserManager in the mocha-hooks.
- create a branching path for the mocha hook to quit the browser instance, so that Playwright's browser.close() function is called instead of Selenium's.

Address code review comments.
- remove commented out code I was keeping for later.
- add file header for `browser-manager.js` to clearly denote this is for Playwright use.

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+/**
```
Should we add a note in the docs or maybe the filename that this is for playwright and not for legacy tests?

worldomonation replied:
Sure, I can do that. Will be coming in next commit.

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+ * By default, this function will return 'desktop' as the target.
+ * To specify another screen size, set the BROWSERSIZE environment variable.
+ *
+ * @returns {string} String representation of the target screen size.
```
I'm curious if Typescript has a place for writing tests. Idk if it's possible, but it might be worth checking out as we do this migration 🤔

worldomonation replied:
It could be a good idea. Playwright has native support for TypeScript, though the limiting factor may be me - I do not know TS :P

noahtallen replied:
Gotcha! That's good to know. If it wasn't too high impact, I almost feel like it would be worth it to create our initial framework and examples using TypeScript -- but that assumes that we _want_ people to write e2e tests in TS long-term. Do we?

cc @sarayourfriend @jsnajdr

sarayourfriend replied:
Even if e2e tests aren't written in TypeScript long-term (I often find writing tests in TypeScript annoying) it's _always_ helpful for test utilities to be strongly and correctly typed (there are some errors in the types applied here). For example, `jest` exports very helpful type definitions that make exploring the library fun and easy for new comers as long as they're using an editor that supports intellisense.

From a "learning TypeScript" perspective, this module is probably a really good example of one that would be not too hard to refactor into TypeScript. I'd be more than happy to pair with you @worldomonation on that if you'd like!

worldomonation replied:
> Even if e2e tests aren't written in TypeScript long-term (I often find writing tests in TypeScript annoying) it's _always_ helpful for test utilities to be strongly and correctly typed (there are some errors in the types applied here).

I am coming new to ES6 (previously a Python developer) so I'm trying to wrap my mind around JS types in general, which is the reason for type mistakes 😆  With that said, I would be happy to refactor `browser-manager.js` to be in TypeScript.

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+		isHeadless = false;
+	}
+
+	// const userAgent = `user-agent=Mozilla/5.0 (wp-e2e-tests) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/${ browser.version() } Safari/537.36`;
```
some commented code here and a few lines below :)

worldomonation replied:
Will be removed in next commit.

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+export async function newBrowserContext() {
+	// If no existing instance of a Browser, then launch a new instance.
+	if ( ! global.__BROWSER__ ) {
+		global.__BROWSER__ = await launchBrowser();
```
Is it best practice in Playwright to use globals to access this stuff?

worldomonation replied:
I modeled the use of Global to store the Browser instance from the current Selenium based suite. I am definitely not certain if this is the best way to manage the browser instance, so I'm open to ideas.

noahtallen replied:
I feel like a local singleton or variable accessed through imports would be more "normal", but I guess test/NodeJS code can be a bit different.

instead of `global.__BROWSER__`, maybe we just do `let browser = null;` in the outer scope of this file, and then either export that or access it with exported functions.

It might also be a good use case for a class, since there is some local "state" to manage. Just some ideas, nothing blocking :)

worldomonation replied:
In the latest iteration I have removed and discontinue usage of `global` objects, instead using module scoped variables that are exported. Thank you for the feedback!

In the future, if it is warranted, it may make sense to extract state-related code represented by the `browser` variable to is own class. At this moment, I don't think it's necessary as the only things that need to be tracked are whether an instance of a browser exists or not.

noahtallen replied:
makes sense

noahtallen reviewed (commented):
Looks solid! Out of curiosity, how do you consume this in tests?

worldomonation commented:
> Looks solid! Out of curiosity, how do you consume this in tests?

This would be similar to how the current `driver-manager.js` is used throughout the test suite:
- provide functions that hook into the `mocha-hooks.js`.
- abstract away the process around the management of browser parameters and setting up necessary environment such as window size, screenshot capture, etc.

I will rebase this PR onto trunk as I have just merged the parent PR to this, #51079.
The example file I have included in this PR isn't really taking advantage of the implementation of the `browser-manager.js` yet. I can make some changes here to illustrate how I imagine future tests will use it.

worldomonation commented:
@noahtallen I've rebased the PR onto the newest `trunk` which should get rid of some of the noise that was present when you reviewed. Notably, changes that were part of the parent commit #51079 were shown in this PR as its own changes as well and that would have been confusing.

I have also addressed issues raised in your comments with a subsequent commit. 

I will make sure to present a cleaner PR next time for review, sorry about that!

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+const browserStartTimeoutMS = 2000;
+
+export let browser;
+export const isPlaywright = true;
```
Is there any need for this, if it is always true and doesn't get changed?

worldomonation replied:
The intention of this value was to be used outside of the `browser-manager` to switch codepaths depending on if Playwright behavior was needed, or legacy Selenium path was needed.

With that said, there has been feedback from Sergio in the follow-on PR (#51209) and he raised an interesting point that `magellan` has the ability to load specific hook files which would remove the need for me to have these flags in the first place.

worldomonation replied:
I've removed the `isPlaywright` flag in preparation for the next patch which will separate out mocha-hooks into Playwright and Selenium versions therefore rendering this flag moot.

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+	let isHeadless = true;
+	if ( ! process.env.HEADLESS || ! config.has( 'headless' ) ) {
+		isHeadless = false;
+	}
```
this could just be:

```suggestion
  const isHeadless = process.env.HEADLESS || config.has( 'headless' );
```

(though, the exact inverse of the current conditional would be `&&`, but I assume we want it if either are true?)

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+	if ( ! process.env.HEADLESS || ! config.has( 'headless' ) ) {
+		isHeadless = false;
+	}
+
```
this could just be:

```suggestion
  const isHeadless = process.env.HEADLESS || config.has( 'headless' );
```

(though, the exact inverse of the current conditional would be `&&`, but I assume we want it if either are true?)

worldomonation replied:
Haha, silly me - thanks for the improvement suggestion! 
The intention is to check if the environment variable is set, or if the config file specifies headless - then take those values, otherwise set it to False.

worldomonation replied:
I have taken your suggestion and modified it to be a ternary expression that explicitly checks if the values set in the environment variable OR config file is true, otherwise default to false. Otherwise, it would complain about expecting a boolean in line 99 but getting a string instead.

noahtallen replied:
@worldomonation Sorry to be nitpicky here again, but I checked locally, and we can should be able to simplify to

```
const const isHeadless = process.env.HEADLESS === 'true' || config.has( 'headless' )
```

According to Typescript, `config.has` should return a boolean. But if does really return a string, then we can at least remove the ternary operator. `config.has( 'headless' ) === 'true' ? true : false` is exactly the same as `config.has( 'headless' ) === 'true'`

worldomonation replied:
Sure, I will change it to what you suggested:
`const isHeadless = process.env.HEADLESS === 'true' || config.has( 'headless' )`

noahtallen commented on test/e2e/lib/browser-manager.js:
```diff
@@ -0,0 +1,119 @@
+ */
+export async function newBrowserContext() {
+	// If no existing instance of a Browser, then launch a new instance.
+	if ( browser === 'undefined' ) {
```
I don't think undefined should be a string here. It might just be worth having `if ( browser == null )` to check for null and undefined at the same time

jsnajdr replied:
The simplest syntax is probably `if ( ! browser )`. `undefined` as string is used typically in `typeof browser === 'undefined'`. That also works if the `browser` variable is not declared at all. Good for checking global variables.

worldomonation replied:
Adopted suggestion in comment with subsequent patch, now it checks for `(! browser)`.

worldomonation commented:
The mobile test on TeamCity is known to be flakier than Desktop as per conversation in #calypso-e2e-task-force.

worldomonation commented:
Tagging team-calypso-platform as I will be going on parental leave for 1 month starting next week, so I would like to merge the stack of PRs into trunk if they pass code review :)

noahtallen commented on test/e2e/specs-playwright/wp-log-in-out-spec.js:
out of curiosity, why don't we need this now?

worldomonation replied:
In the next patch stack (which needs to be rebased on this branch after TS conversion), I connect the `after` hook to handle the browser closure. This way, the test writer doesn't need to be concerned with boilerplate code to handle browser closures.

noahtallen replied:
cool!

worldomonation committed:
Experimental migration to TypeScript for browser-manager.ts.

- removed browser-manager.js from tracked file.
- added tscnfig.json modelled primarily after one found in `client`.
- added some dependency packages in package.json.
- added `yarn build` script.

Lint files

Fix issues with tsconfig file.
- toggle some of the options such as noEmit such that output JS files are written and can be used by the tests.
- restore the after hook.

worldomonation commented:
@noahtallen @sarayourfriend I've attempted a migration of the `browser-manager.js` to a TypeScript file.

As I am new to TS migration please let me know if I have done something that does not make sense, or could be done better.

I have implemented this flow for the test files under `specs-playwright/`:
1. call `yarn build` to compile TS to JS. This would produce output files in JS that would then be imported by the tests.
2. run the tests.

[It appears to run as expected](https://teamcity.a8c.com/buildConfiguration/calypso_RunCalypsoPlaywrightE2eTests/6033312?), which I take as a good sign.

noahtallen commented on test/e2e/lib/browser-manager.ts:
```diff
@@ -0,0 +1,116 @@
+ *
+ * @returns {string} String representation of the target screen size.
+ */
+export function targetScreenSize(): string {
```
If we know what the other screen sizes are allowed to be, it might be nice to specify them inline like `targetScreenSize(): 'desktop' | 'mobile'`

This might also be useful for targetLocale (perhaps a type already exists for those)

worldomonation replied:
I suppose this comment is covering [this section](https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/typescript.md#usage-patterns) in the TypeScript guidelines?

In this case, the return value defaults to `desktop`, but it could be any of `laptop, mobile, desktop or tablet`. How should I handle this case - would it be better to list all four expected outputs as the return type?

The same would go for locale as well. It defaults to `en`, but it could be any one of the multiple strings we support.

worldomonation replied:
I've pushed a new revision that removes usage of primitives for the `screenSize` and `localeCode` related functions. Can you take a look and let me know if I'm on the right path, thanks!

worldomonation committed:
Do not use primitive types for screenSize and localeCode.
Also rename the functions to use action words ie. get

scinos commented on test/e2e/package.json:
```diff
@@ -67,6 +69,7 @@
 		"testarmada-magellan": "11.0.10",
 		"testarmada-magellan-local-executor": "^2.0.3",
 		"testarmada-magellan-mocha-plugin": "github:Automattic/magellan-mocha-plugin#b731cdf6f27a6fa6e45ddcba0b8741370b2df523",
+		"typescript": "^4.1.3",
```
In our repo only the root `package.json` depends on `typescript`, no need to add it here.

scinos commented on test/e2e/tsconfig.json:
```diff
@@ -0,0 +1,22 @@
+{
```
I'd suggest this tsconfig file, more similar to what we have in other places (eg `packages/*/tscofig.json`) but adapted for Node:

```
{
	"compilerOptions": {
		"target": "es2020",
		"lib": [ "DOM", "es2020" ],
		"baseUrl": ".",
		"module": "commonjs",
		"allowJs": false,
		"declaration": true,
		"declarationDir": "types",
		"outDir": "dist",
		"rootDir": ".",

		"strict": true,
		"noUnusedLocals": true,
		"noUnusedParameters": true,
		"noImplicitReturns": true,
		"noFallthroughCasesInSwitch": true,

		"moduleResolution": "node",
		"esModuleInterop": true,

		"forceConsistentCasingInFileNames": true,

		"types": [ "node" ],

		"noEmitHelpers": true,
		"importHelpers": true,

		"composite": true
	},
	"include": [ "lib" ],
}
```

It will output the compiled files to `test/e2e/dist`, which I find a bit cleaner that saving them in the same dir than the `.ts` files.

worldomonation replied:
Thanks. I only knew enough to get into trouble with my configuration, so I appreciate your suggestion.

scinos commented on test/e2e/specs-playwright/wp-log-in-out-spec.js:
```diff
@@ -1,32 +1,50 @@
+/**
+ * Internal dependencies
+ */
+import * as browserManager from '../lib/browser-manager';
```
If you use the `tsconfig.json` that I proposed below, this should be changed to `../dist/lib/browser-manager`

scinos commented on test/e2e/package.json:
```diff
@@ -10,6 +10,7 @@
 	"main": "greetings.js",
 	"defaultTestArgs": "-g -s $BROWSERSIZE",
 	"scripts": {
+		"build": "tsc --build ./tsconfig.json",
```
To keep it consistent with other TS packages in the repo, I'd add

```
		"clean": "tsc --build ./tsconfig.json --clean && npx rimraf dist",
```

worldomonation replied:
Will be included in patch to address your comments.

scinos commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -766,6 +766,9 @@ object RunCalypsoPlaywrightE2eTests : BuildType({
 				openssl aes-256-cbc -md sha1 -d -in ./config/encrypted.enc -out ./config/local-test.json -k "%CONFIG_E2E_ENCRYPTION_KEY%"
 
+				# Compile TypeScript
+				yarn build
```
nit: This should be part of `"Prepare environment"`

worldomonation replied:
You're right. Will move it in next commit.

worldomonation committed:
Address Sergio's code review comments.

worldomonation commented:
Rebased onto recent trunk, hopefully more of the tasks pass (for peace of mind) before merging to trunk.

worldomonation committed:
Add TypeScript output to ignore files.

sarayourfriend commented on test/e2e/specs-playwright/wp-log-in-out-spec.js:
```diff
@@ -1,32 +1,50 @@
+/**
+ * Internal dependencies
+ */
+import * as browserManager from '../dist/lib/browser-manager';
```
Was this supposed to import from `dist` or directly from `./lib/browser-manager`?

worldomonation replied:
My understanding (correct me if I'm wrong) is that plain JS file cannot import TS. I implemented a workflow that requires compilation to JS which is output into the `dist/` directory, and that is imported in the tests.

sarayourfriend replied:
Ah, plain JS should be able to import TypeScript just fine as we use the babel parser for TypeScript. See https://github.com/automattic/wp-calypso/blob/907b61456eab841f0941e5ee1e4f302b339d0ecb/client/sections-middleware.js#L18 for an example (`add-reducer` is TypeScript).

scinos replied:
This is meant to be run by Node. I think Mocha uses [esm](https://github.com/standard-things/esm) to bring in support for ESM imports without transpiling the code first with babel.

At least in my local tests this has to point to `../dist/lib/browser-manager`, otherwise Mocha+esm won't be able to find it.

worldomonation replied:
@scinos I have a follow-up PR where I was able to load TypeScript into plain JavaScript test files using `ts-node`, if you'd like to take a look. https://github.com/Automattic/wp-calypso/pull/51419

sarayourfriend commented on test/e2e/lib/browser-manager.ts:
```diff
@@ -0,0 +1,124 @@
+/**
+ * External dependencies
+ */
+import { Browser, BrowserContext, chromium } from 'playwright';
```
Just a nit but it's good practice to separate type imports into their own `import type` statements: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

worldomonation replied:
Got it, I will change this in the next patch in the stack.

WunderBart commented on test/e2e/lib/browser-manager.ts:
```diff
@@ -0,0 +1,124 @@
+ * @returns {void} No return value.
+ */
+export function quitBrowser(): void {
+	browser.close();
```
Because `browser.close()` returns a promise, I think it would be best to `return` this line so we can properly `await` for it to resolve.

worldomonation replied:
According to the [documentation](https://playwright.dev/docs/api/class-browser#browserclose) it does not return anything, I think?

WunderBart replied:
Looks like it does return a promise according to https://github.com/microsoft/playwright/blob/master/src/client/browser.ts#L79. In the examples here you can see it's being awaited for https://github.com/microsoft/playwright#page-screenshot.

worldomonation replied:
Argh void promise. Thanks for point that out!

WunderBart commented on test/e2e/lib/browser-manager.ts:
```diff
@@ -0,0 +1,124 @@
+
+	// Generate a new BrowserContext.
+	return await browser.newContext( {
+		viewport: null, // Do not override window size set in the browser launch parameters.
```
I'm wondering if we need this extra abstract layer for the browser instance. Is it intended that we don't want to return the current browser instance? What if we only had the `launchBrowser` method that returns the browser instance and dropped the `quitBrowser` and `newBrowserContext` in favor of the native ones? I guess it would require one extra step to get the `context` comparing to the current flow, but at the same time, we're not adding an extra learning curve to the Playwright framework. So, in a spec it could look like this:

```js
import { launchBrowser } from 'browser-manager';

let browser;
let context;

before() {
	browser = await launchBrowser();
	context = await browser.newContext();
}

after() {
	browser.close();
}

describe('Howdy', function() {
	// write some specs
});

```

What do you think, folks?

WunderBart replied:
Ah, now I can see that we're sharing the browser instance with Mocha [root hooks](https://github.com/Automattic/wp-calypso/pull/51285/files#diff-5580bcd7dfbb3f4ef83c4299ea79d943d2175fc3fd08b56abdd49808812fbd13R24). Would it make sense then to make the `browser` instance global? (i.e. Launch in `before` and close in `after` root hook)

worldomonation replied:
Yes, my intention was to use browser instance in mocha hooks to control browser launch and close. The overall goal I was looking to achieve is to reduce the amount of boilerplate code that test writers need to include in each spec.

With that said however, in my subsequent patches I am discovering that the goal might not be feasible, so this is subject to change.

WunderBart commented:
@worldomonation sorry for commenting post-merge - didn't get to this one in time!
